src/main/java/com/google/sps/data/RestrictTimesCriteria.java

```diff
@@ -0,0 +1,53 @@
+package com.google.sps.data;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Arrays;
+import java.util.ArrayList;
+
+public class RestrictTimesCriteria implements Preference {
+  public List<TimeRange> userNoClassTimes;
+  
+  public RestrictTimesCriteria(List<TimeRange> userNoClassTimes) {
+    this.userNoClassTimes = userNoClassTimes;
+  }
+
+  /**
+   * Implements the preference score of the schedule, based on the criteria
+   * that prioritizes the TimeRanges that a user does not want class during
+   */
+  public float preferenceScore(Schedule schedule) {
+    // count the number of minutes that overlap
+    float minutesPenalty = 0;
+    
+    List<Course> courses = (List) schedule.getCourses();
+    List<TimeRange> timeOrderedCourses = getCourseTimes(courses);
+
+    for (TimeRange event : userNoClassTimes) {
+      for (TimeRange course : timeOrderedCourses) {
+        if (event.overlaps(course)) {
+          float timeDiff = (float) calculateDifference(event, course);
+          minutesPenalty -= timeDiff;
+        }
+      }
+    }
+    return minutesPenalty;
+  }
+
+  // Calculates the overlapping minutes between two TimeRanges
+  int calculateDifference(TimeRange a, TimeRange b) {
+    int start = Math.max(a.start(), b.start());
+    int end = Math.min(a.end(), b.end());
+    return end - start;
+  }
+
+  // Returns a List of TimeRanges from a list of courses
+  List<TimeRange> getCourseTimes(List<Course> courses) {
+    List<TimeRange> times = new ArrayList<TimeRange>();
+    for(Course course : courses) {
+      times.addAll(course.getSections().get(0).getMeetingTimes());
```

Comment:
`get(0)` ?

Can you assume this is always non-empty? Do you always want to ignore elements beyond the first?

Reply:
This is the way that the `Scheduler::generateSchedules()` function outputs it's schedules. Each course in the outputted schedule has only one section (in accordance to the invariants, which state that no schedule can have multiple course sections in it).

Reply:
As Maceo mentioned, each course in a schedule will only have one section (an invariant), so it the course section is always the first element of the sections list. Perhaps I can discuss with Maceo on a more clear approach, i.e. `getSection()` that only returns that section, and avoids the confusing `getSections().get(0)`

Reply:
Ok. I left a comment about this from the `Scheduler::generateSchedules()` point of view on another pull request here: https://github.com/googleinterns/step78-2020/pull/15#discussion_r451746758

Let me know how you would like to resolve this between both pull requests.

The easiest option, if you decide to go with the approach suggested in the linked comment, would be to open up a separate pull request to add `CourseEnrollment` and update the `Scheduler` API accordingly (only the return types and minimum required changes to make the types check out). Get that approved and merged quickly, then update both this and PR #15 accordingly.

...or you can do it gladiator style and see who gets to merge first then deal with the ensuing merge conflicts & breakages. Let me know what you settle on.

Reply:
See https://github.com/googleinterns/step78-2020/pull/15#discussion_r451785869

Reply:
We decided on creating the `ScheduledCourse` class, representing a course that's added to the schedule with a single section rather than a `List<Section>`

Reply:
I will make the changes to change `getSections().get(0)` once [PR  #21](https://github.com/googleinterns/step78-2020/pull/21) gets merged in.

Reply:
21 is merged!